### ISSUE TITLE
Tokens getting replaced by <unk> even when the threshold is 0

### ISSUE THREAD

**Issue by riteshpanjwani** [facebookresearch/fairseq-lua#35]
I ran the following for Data Pre-processing:

```
fairseq preprocess -sourcelang de -targetlang en \
  -trainpref $TEXT/train -validpref $TEXT/valid -testpref $TEXT/test \
  -thresholdsrc 0 -thresholdtgt 0 -destdir data-bin/iwslt14.tokenized.de-en
```

As you can see, I set `-thresholdsrc` & `-thresholdtgt` to **0**, still it is replacing tokens in validation and test set with the **unknown** symbol.

Here's the output of the above command:

```
| [de] Dictionary: 113533 types
| [de] data/iwslt14.tokenized.de-en/train.de: 160215 sents, 3258560 tokens, 0.00% replaced by <unk>
| [de] data/iwslt14.tokenized.de-en/valid.de: 7282 sents, 149250 tokens, 1.91% replaced by <unk>
| [de] data/iwslt14.tokenized.de-en/test.de: 6750 sents, 132488 tokens, 2.45% replaced by <unk>
| [de] Wrote preprocessed data to data-bin/iwslt14.tokenized.de-en
| [en] Dictionary: 53330 types
| [en] data/iwslt14.tokenized.de-en/train.en: 160215 sents, 3433520 tokens, 0.00% replaced by <unk>
| [en] data/iwslt14.tokenized.de-en/valid.en: 7282 sents, 157272 tokens, 0.59% replaced by <unk>
| [en] data/iwslt14.tokenized.de-en/test.en: 6750 sents, 137891 tokens, 0.85% replaced by <unk>
| [en] Wrote preprocessed data to data-bin/iwslt14.tokenized.de-en
```

In validation & test data, value should also be **0.00 %** right? I am using this for some other task where I don't want the tokens to be replaced by **unknown**

Is there some option I am missing here?

**Comment by jgehring** [comment]
The dictionary is built from the training set only. In this case, the validation and test sets contain words that are not in the training set and those will need to be replaced by `<unk>`. If you want to make sure that there are absolutely no unknown words in the validation and test set you'll need to have a backup strategy for these words, e.g. by training and applying a BPE vocabulary before pre-processing.

**Comment by riteshpanjwani** [comment]
@jgehring Okay, but I am not using this library for NMT & most of the words in my validation & test set are not present in the training set. Is there anyway I can bypass this replacement by changing the code? Can you please point me, where & how to change it?
Or can you please elaborate this:
"by training and applying a BPE vocabulary before pre-processing."
Thanks.

**Comment by jgehring** [comment]
What do you mean by "bypass this replacement"? What do you want the pre-processing to do if it encounters an unknown word in the valid or train set? If you want to add it to the dictionary, it would be the same as replacing it by an unknown word -- the word is not present in the training set and thus its embedding will not be trained. As a result, the model will never be able to make any sense of this word.

Regarding BPE, you can use https://github.com/rsennrich/subword-nmt to train and a apply a sub-word vocabulary before pre-processing. For more information, check out the corresponding paper [here](http://www.aclweb.org/anthology/P16-1162).

**Comment by jgehring** [comment]
@riteshpanjwani, do you have any further questions on this?
